### PR TITLE
[Fix][Transform-V2] Fix datetime/numeric/string function edge cases and update tests

### PR DESCRIPTION
**Pull request by yzeng1618**
https://github.com/apache/seatunnel/issues/10137

### Purpose of this pull request

Fix several edge cases of SQL transform built‑in functions (DATEDIFF with MONTH, WEEK, SIGN, ASCII/LEFT/RIGHT, and CAST), and update corresponding unit tests and E2E expectations.

### Does this PR introduce _any_ user-facing change?
Yes.

- DATEDIFF(..., 'MONTH') now returns the total month difference across years (e.g. from 2023-01-01 to 2024-03-01 returns 14 instead of 15).

- WEEK() uses the ISO week number without the previous extra +1.

- SIGN(null) now returns NULL instead of 0.

- ASCII(NULL) / ASCII('') now return NULL instead of throwing, and LEFT / RIGHT with negative length return an empty string.

- CAST no longer returns NULL when the value text equals the type name (e.g. CAST('VARCHAR' AS VARCHAR) now returns "VARCHAR").

These behavior changes only affect the dev branch and fix previously incorrect or inconsistent results.

### How was this patch tested?

Updated and added unit tests in seatunnel-transforms-v2:

- DateTimeFunctionTest for DATEDIFF month across years.

- NumericFunctionTest for round on SHORT and SIGN(NULL).

- StringFunctionTest for ASCII(NULL/'') and LEFT/RIGHT with negative length.

- SystemFunctionTest for CAST and COALESCE with target type.

- Updated the datetime E2E config in seatunnel-e2e to align with the corrected DATEDIFF behavior.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If necessary, please update `incompatible-changes.md` to describe the incompatibility caused by this PR.
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)

### CODE REVIEW

File: seatunnel-transforms-v2/src/test/java/org/apache/seatunnel/transform/sql/zeta/NumericFunctionTest.java [hunk 1]
```diff
@@ -78,4 +79,18 @@ public void testTrimScale() {
         Assertions.assertEquals("0", NumericFunction.trimScale(Collections.singletonList(0)));
         Assertions.assertNull(NumericFunction.trimScale(Collections.singletonList((Object) null)));
     }
+
+    @Test
+    public void testRoundShortNegativeScale() {
+        short shortValue = 123;
+
+        Number result = NumericFunction.round(Arrays.asList(shortValue, -1));
+
```
Comment (chl-wxp) [inline seatunnel-transforms-v2/src/test/java/org/apache/seatunnel/transform/sql/zeta/NumericFunctionTest.java]: Remove a lot of newlines in your code like this

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/functions/NumericFunction.java [hunk 2]
```diff
@@ -208,7 +208,7 @@ private static Number round(Number v1, Number v2, RoundingMode roundingMode) {
         c:
         switch (t.toUpperCase()) {
             case "INTEGER":
-            case "SHOT":
+            case "SHORT":
             case "LONG":
                 {
```
Comment (chl-wxp) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/functions/NumericFunction.java]: case "INTEGER":
case "SHORT":
case "LONG": 
...
Can the hard coding here be optimized into enumeration writing?

File: seatunnel-transforms-v2/src/test/java/org/apache/seatunnel/transform/sql/zeta/DateTimeFunctionTest.java [hunk 3]
```diff
@@ -176,4 +179,14 @@ public void testFromUnixtimeFunctionWithIntegerInput() {
...
+
+    @Test
+    public void testDateDiffMonthAcrossYearUsesTotalMonths() {
+        LocalDate start = LocalDate.of(2023, 1, 1);
+        LocalDate end = LocalDate.of(2024, 3, 1);
+
+        Long months = DateTimeFunction.datediff(Arrays.asList(start, end, "MONTH"));
+
+        Assertions.assertEquals(14L, months);
+    }
```
Comment (zhangshenghang) [inline seatunnel-transforms-v2/src/test/java/org/apache/seatunnel/transform/sql/zeta/DateTimeFunctionTest.java:191]: Please add this change record at https://github.com/apache/seatunnel/blob/dev/docs/en/concept/incompatible-changes.md

Comment (yzeng1618) [reply seatunnel-transforms-v2/src/test/java/org/apache/seatunnel/transform/sql/zeta/DateTimeFunctionTest.java:191]: Has been added

**Review by corgy-w — APPROVED** [review]

**Review by Carl-Zhou-CN — APPROVED** [review]
+1
